src/libstore/binary-cache-store.cc

```diff
@@ -162,23 +155,26 @@ void BinaryCacheStore::addToStore(const ValidPathInfo & info, Source & narSource
     /* Read the NAR simultaneously into a CompressionSink+FileSink (to
        write the compressed NAR to disk), into a HashSink (to get the
        NAR hash), and into a NarAccessor (to get the NAR listing). */
-    HashSink fileHashSink(htSHA256);
+    HashSink fileHashSink { htSHA256 };
     std::shared_ptr<FSAccessor> narAccessor;
+    HashSink narHashSink { htSHA256 };
     {
     FdSink fileSink(fdTemp.get());
```

Comment:
This is beyond the scope of this PR, but since we're revisiting these store API methods it's worth noting that we could optimize away the tmpfile if we know a little bit more in advance. Although optimizing away a tmpfile seems unimpressive, it changes the time taken from `O(sum(steps))` to `O(max(steps))`, which is significant when compression and upload take similar amounts of time.

***When the file is known to be small (low hanging fruit)***

Add a size parameter to `addToStoreCommon` or use a fancy sink that only writes to file when it crosses a limit. This is similar to what `LocalStore::addToStoreFromDump` does.

***When we know the nar hash in advance***

For http binary caches this does require us to change the binary cache filenames to match *uncompressed* hashes, which seems to be equivalent and can only result in one-time duplication in existing caches when new paths are uploaded.
I don't know yet how IPFS caches fit into this picture, but if those can compress after hashing, this would be beneficial.
Another reason to do this is so we don't need to compress before we can decide to reuse an available nar file.

In this case it does make sense to have both `addToStore(const ValidPathInfo & info, .....)` and `addToStore(....., std::function<ValidPathInfo(HashResult))` where the prior can have a default implementation in terms of the latter.

Reply:
I decided not to try to optimize away the temporary file for small NARs because the overhead is likely to be insignificant compared to stuff like HTTP requests.